README.md
[![Build Status](https://travis-ci.org/stewid/gix.png)](https://travis-ci.org/stewid/gix)

# Introduction

`gix` is an R package that aims to index and search a Git repository
using the [Xapian](http://xapian.org/) search engine.

## Index a repository

Index the content of a git repository to documents with the Xapian
search engine. A `document` is the data returned from a search. The
indexing creates a database located in the `.xapian` folder in the
repository.


```r
library(gix)

## Index the 'gix' repository. (Current directory)
gixit(".", github = "stewid/gix")
```

## Search a repository


```r
## Search for author Widgren and display url to blob on GitHub
gix(".", "author:widgren")$url
```

```
##  [1] "https://github.com/stewid/gix/blob/c6fc2f018ee81cf973495dc359995cdac6fb4641/DESCRIPTION"
##  [2] "https://github.com/stewid/gix/blob/54f3f6f422a352184caf115dabe8567ed7295fd0/.Rbuildignore"
##  [3] "https://github.com/stewid/gix/blob/54f3f6f422a352184caf115dabe8567ed7295fd0/NAMESPACE"
##  [4] "https://github.com/stewid/gix/blob/54f3f6f422a352184caf115dabe8567ed7295fd0/NEWS"
##  [5] "https://github.com/stewid/gix/blob/54f3f6f422a352184caf115dabe8567ed7295fd0/README.md"
##  [6] "https://github.com/stewid/gix/blob/54f3f6f422a352184caf115dabe8567ed7295fd0/man/gix.Rd"
##  [7] "https://github.com/stewid/gix/blob/8906a4a3ddbb7cd29e33d47b37cdbf0f5ab70c2a/DESCRIPTION"
##  [8] "https://github.com/stewid/gix/blob/a489b0a4272be35043967b42e5caa6ed1b66ba26/.gitignore"
##  [9] "https://github.com/stewid/gix/blob/eca0aa4931fd0d6547841a7737fd3f4ec728f9d9/man/gixit-methods.Rd"
## [10] "https://github.com/stewid/gix/blob/7e5bac365cad12fc5a5e9cbd8f4030bb0ebb3031/DESCRIPTION"
```


```r
## Search for sha that starts with 'd7'
gix(".", "sha:d7*")[, c("sha", "url")]
```

```
##                                        sha
## 1 d7f105139782ab695d86613e343916f7372f4ac0
##                                                                                   url
## 1 https://github.com/stewid/gix/blob/1f25d61fff69361e748efbcf07b87846d81ecd1d/LICENSE
```


```r
## Search for 'biarch' and view date of author commit and url
gix(".", "biarch")[, c("when", "url")]
```

```
##                   when
## 2  2014-09-13 23:03:38
## 10 2014-09-13 23:27:32
## 16 2014-09-18 18:44:32
## 24 2014-09-20 14:40:54
## 32 2014-10-01 14:18:34
##                                                                                        url
## 2  https://github.com/stewid/gix/blob/c6fc2f018ee81cf973495dc359995cdac6fb4641/DESCRIPTION
## 10 https://github.com/stewid/gix/blob/8906a4a3ddbb7cd29e33d47b37cdbf0f5ab70c2a/DESCRIPTION
## 16 https://github.com/stewid/gix/blob/7e5bac365cad12fc5a5e9cbd8f4030bb0ebb3031/DESCRIPTION
## 24 https://github.com/stewid/gix/blob/e7b9a4fd794bb1dd408f95db39a7eb461825122e/DESCRIPTION
## 32 https://github.com/stewid/gix/blob/7d0970e2a2304abb92ddbd2f6cdf583e2ad31708/DESCRIPTION
```

## Installation

The following two R packages must be installed:

 - [xapr](https://github.com/stewid/xapr)
 - [git2r](https://github.com/ropensci/git2r)

To install the development version of `gix`, it's easiest to use the
devtools package:


```r
# install.packages("devtools")
library(devtools)
install_github("stewid/gix")
```
Another alternative is to use `git` and `make`

```
$ git clone https://github.com/stewid/gix.git
$ cd gix
$ make install
```

**NOTE:** The package is in a very early development phase. Functions
and documentation may be incomplete and subject to
change. Suggestions, bugs, forks and pull requests are
appreciated. Get in touch.

# License

The `gix` package is licensed under the GPL (>= 2). See these files
for additional details:

- [LICENSE](LICENSE)     - `gix` package license


NEWS
gix 0.1
-------

NEW FEATURES

* gixit - method to index a git repository

* gix - method to search a git repository


gix 0.0.1
-----------

NEW FEATURES

* Initial package structure


R/gix.r
## gix, search a Git repository using the Xapian search engine.
## Copyright (C) 2014 Stefan Widgren
##
##  gix is free software; you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation; either version 2 of the License, or
##  (at your option) any later version.
##
##  gix is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License along
##  with this program; if not, write to the Free Software Foundation, Inc.,
##  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

##' Search a git repository
##'
##' @rdname gix-methods
##' @docType methods
##' @param repo The repository
##' @param query The search query
##' @return :TODO:DOCUMENTATION:
##' @import git2r
##' @keywords methods
setGeneric("gix",
           signature = c("repo", "query"),
           function(repo, query)
           standardGeneric("gix"))

##' @rdname gix-methods
##' @export
setMethod("gix",
          signature(repo  = "character",
                    query = "character"),
          function(repo, query)
          {
              gix(repository(repo), query)
          }
)

##' @rdname gix-methods
##' @export
setMethod("gix",
          signature(repo  = "git_repository",
                    query = "character"),
          function(repo, query)
          {
              db <- new("xapian_database",
                        path = file.path(workdir(repo), ".xapian"))

              xsearch(
                  query    = query,
                  db = db,
                  . ~ author:A + sha:XSHA,
                  wildcard = TRUE)
          }
)


R/gix-package.r
## gix, search a Git repository using the Xapian search engine.
## Copyright (C) 2014 Stefan Widgren
##
##  gix is free software; you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation; either version 2 of the License, or
##  (at your option) any later version.
##
##  gix is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License along
##  with this program; if not, write to the Free Software Foundation, Inc.,
##  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

##' Search a Git repository
##'
##' gix is a package to index and search a Git repository using the
##' Xapian search engine.
##' @docType package
##' @name gix
##' @import git2r
##' @import xapr
##' @include gix.r
NULL


R/gixit.r
## gix, search a Git repository using the Xapian search engine.
## Copyright (C) 2014 Stefan Widgren
##
##  gix is free software; you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation; either version 2 of the License, or
##  (at your option) any later version.
##
##  gix is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License along
##  with this program; if not, write to the Free Software Foundation, Inc.,
##  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

##' Index a git repository
##'
##' @rdname gixit-methods
##' @docType methods
##' @param path Path to the repository
##' @param ... Additional arguments affecting the indexing
##' ##' @return invisible(NULL)
##' @keywords methods
setGeneric("gixit",
           signature = "path",
           function(path, ...)
           standardGeneric("gixit"))

##' @rdname gixit-methods
##' @param callbacks A list of callback functions to enable more
##' control over how blob's are indexed. Each list item, named by file
##' extension, is a function with the two arguments 'content' and
##' 'sha'. Note: not implemented yet.
##' @param github Optional repository address in the format
##' 'username/repo'. Generates a link to the content on github in the
##' document area. Default is NULL.
##' @return invisible NULL
##' @export
setMethod("gixit",
          signature(path = "character"),
          function(path,
                   callbacks = NULL,
                   github    = NULL)
          {
              repo <- repository(path)

              ## List all objects in the repository
              blobs <- odb_blobs(repo)

              ## :TODO:FIXME: Design a default index plan. Blobs might
              ## contain duplicate entries for each sha with different
              ## path and name. For now, index only the content of
              ## unique sha's.
              blobs <- blobs[!duplicated(blobs$sha),]

              blobs$content <- sapply(seq_len(nrow(blobs)), function(i) {
                  content(lookup(repo, blobs$sha[i]), split = FALSE)
              })

              if (!is.null(github)) {
                  ## Create a link to the content on github
                  blobs$url <- paste0("https://github.com/",
                                      github,
                                      "/blob/",
                                      blobs$commit,
                                      ifelse(nchar(blobs$path) > 0,
                                             paste0("/", blobs$path, "/"),
                                             "/"),
                                      blobs$name)
              }

              xindex(. - content ~ X*path + X*sha + A:author + Q:sha + content,
                     blobs,
                     file.path(path, ".xapian"))

              invisible(NULL)
          }
)
